example/mmap.cc

```diff
@@ -39,46 +39,94 @@ auto open_mem_file(int pages) -> int {
+  size_t alloc_total_size;
   int fd;
 };
 
 auto make_mem(size_t size, int fd) -> std::unique_ptr<mem_info> {
   std::cout << "> Making memory "
-    << "(size = " << size << ", fd = " << fd << ")..." << std::endl;
-  auto offset = wasm::v8::Memory::redzone_size_lo(size);
-  auto alloc_size = size + wasm::v8::Memory::redzone_size_hi(size);
-
-  void* base = nullptr;
-  if (offset > 0) {
-    // Lo redzone needed.
-    base = mmap(nullptr, offset, PROT_NONE, MAP_ANON | MAP_PRIVATE, 0, 0);
-    if (base == MAP_FAILED) {
-      std::cout << "> Error reserving lo redzone! errno = " << errno
-        << " (" << strerror(errno) << ")" << std::endl;
-      exit(1);
-    }
+            << "(size = " << size << ", fd = " << fd << ")..." << std::endl;
+  // How much memory we need to reserve including guard regions before and after
+  auto alloc_total_size = wasm::v8::Memory::redzone_size_lo(size) + size +
+                          wasm::v8::Memory::redzone_size_hi(size);
+  std::cout
+      << std::hex
+      << "> linux_page_size = 0x"
+      << getpagesize()
+      << ", wasm_page_size = 0x"
+      << wasm::Memory::page_size
+      << ", memory_size = 0x"
+      << size
+      << ", redzone_size_lo = 0x"
+      << wasm::v8::Memory::redzone_size_lo(size)
+      << ", redzone_size_hi = 0x"
+      << wasm::v8::Memory::redzone_size_hi(size)
+      << ", total_size = 0x"
+      << alloc_total_size
+      << std::dec
+      << std::endl;
+
+  std::cout
+      << "> Calling mmap "
+      << "(addr = nullptr"
+      << ", size = "
+      << alloc_total_size
+      << ")..."
+      << std::endl;
+
+  // Need to allocate the total address space first and only then remap a
+  // portion of it with MAP_FIXED
+  void *base =
+      mmap(nullptr, alloc_total_size, PROT_NONE, MAP_ANON | MAP_PRIVATE, 0, 0);
+  if (base == MAP_FAILED) {
+    std::cout << "> Error reserving lo redzone! errno = " << errno << " ("
+              << strerror(errno) << ")" << std::endl;
+    exit(1);
   }
 
-  auto data = static_cast<byte_t*>(base) + offset;
-  auto result = mmap(
-    data, alloc_size, PROT_NONE, MAP_FILE | MAP_FIXED | MAP_SHARED, fd, 0);
+  // Where the module data is located
+  auto data =
+      static_cast<byte_t *>(base) + wasm::v8::Memory::redzone_size_lo(size);
+
+  std::cout
+      << "> base_address = "
+      << static_cast<void *>(base)
+      << ", data_address = "
+      << static_cast<void *>(data)
+      << std::endl;
+
+  std::cout
+      << "> Calling mmap "
+      << "(addr = "
+      << static_cast<void *>(data)
+      << ", size = "
+      << size
+      << ", fd = "
+      << fd
+      << ")..."
+      << std::endl;
+
+  // Modify the existing mapping such that `size` bytes beginning at `base +
+  // offset` are mapped with the contents of the file descriptor `fd`. We remap
+  // the maximum amount of memory a module can use but only make the `size`
+  // bytes accessible.
+  auto result = mmap(data, size + wasm::v8::Memory::redzone_size_hi(size),
```

Comment:
This is fine; it doesn't affect the size of the file. From `man mmap`:

> The mmap() function can be used to map a region of memory that is larger than the current size of  the  object. Memory access within the mapping but beyond the current end of the underlying objects may result in SIGBUS signals being sent to the process.

Digging through the docs a bit more, perhaps `grow_mem` should either call `mmap` again or maybe `mremap` after extending the file?

> If the size of the mapped file changes after the call to mmap() as a result of  some  other  operation  on  the mapped file, the effect of references to portions of the mapped region that correspond to added or removed portions of the file is unspecified.

`grow_mem` might be OK because we never access these regions before growing the file. This certainly works but I'm not sure it is 100% safe.

Reply:
> This is fine; it doesn't affect the size of the file. From man mmap:

I don't follow. The file created in the example has 128K (2 pages), while IIUC, the mapping you changed to would require it to be 2G+128K, to include the lower end redzone; grow_mem makes a similar assumption. I'm a bit confused why this version works at all, I'd expect all actual memory accesses to sigbus?

The file should not include the redzone, that would be very wasteful. I'm not an mmap wizard, is there a way to remap just the *middle* of an already mapped region?

> grow_mem might be OK because we never access these regions before growing the file. This certainly works but I'm not sure it is 100% safe.

Yeah, "unspecified" doesn't sound good. We need to make sure that both valid accesses and invalid accesses (beyond the memory's end) behave in a well-defined way, and exactly as required.

Reply:
> The file created in the example has 128K (2 pages)

That is still the case.

>  I'm not an mmap wizard, is there a way to remap just the middle of an already mapped region?

That is what is happening. The first `mmap` call:
```c++
void *base =
      mmap(nullptr, alloc_total_size, PROT_NONE, MAP_ANON | MAP_PRIVATE, 0, 0);
```
Registers all the virtual memory we will need:
```c++
  auto alloc_total_size = wasm::v8::Memory::redzone_size_lo(size) + size +
                          wasm::v8::Memory::redzone_size_hi(size);
```
Then the offset `data` is the part of that address space just after the lower end redzone:
```c++
  auto data =
      static_cast<byte_t *>(base) + wasm::v8::Memory::redzone_size_lo(size);
```
And we make another `mmap` call with `MAP_FIXED` to modify that mapping, beginning at `data` (i.e. `base` + LO), to be backed by a file:
```c++
  auto result = mmap(data, size + wasm::v8::Memory::redzone_size_hi(size),
                     PROT_NONE, MAP_FIXED | MAP_SHARED, fd, 0);
```
After this we have two mappings:
1. `base .. base + LO` not backed by a file
2. `base + LO .. HI` backed by `fd`

And the size of the file pointed to by `fd` has not changed and is still 128K.

Both mappings so far are `PROT_NONE` so any access will end in SIGBUS. Finally we only make `data .. data + size` region accessible:
```
  if (mprotect(data, size, PROT_READ | PROT_WRITE) != 0) {
```
Which matches the size of the memory file.

Since we are going to grow the Wasm module memory and the backing file over time I went with mapping `size + HI` (the maximum module can have, right?) so we can make more memory accessible by first growing the file and then changing the access protection. The address space is already reserved and because the memory was made inaccessible there are no references from the mapped region to the file.

I think this is perfectly valid use of `mmap`.

The alternative would be to create a new mapping every time we grow the memory:

```c++
  // 1. allocate entire address space
  auto base = mmap(nullptr, alloc_total_size, ...);

  // 2. map the initial memory
  auto data = mmap(base + LO, init_size, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_SHARED, fd, 0);

  // to later grow the memory...

  // resize the file
  ftruncate(fd, new_size);
  // 3. create a new mapping that only maps the newly added portion of the file at the end of existing mapping `data + old_size`:
  auto result = mmap(data + old_size, new_size - old_size, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_SHARED, fd, old_size);
```
The difference here is that we create a new mapping in the kernel for each chunk of memory we grow. After step `3.` there will be 4 mappings:
```
---------------------------------------------------
| LO redzone | init_size | new_chunk | HI redzone |
---------------------------------------------------
^            ^           ^           ^
^            ^           ^           no backing
no backing   ^           fd@[init_size .. new_size)
             ^
             fd@[0..init_size)
```

Reply:
Ah, sorry, I somehow managed to miss that `data` was pointing into the middle. Thanks for the rundown. One point, though:

> Since we are going to grow the Wasm module memory and the backing file over time I went with mapping size + HI (the maximum module can have, right?)

No, the maximum memory size is 4G. The size of the hi redzone is platform-specific, and may range from 0 (e.g., on 32 bit) to something much larger than 4G (on 64 bit, size + hi_redzone_size = 8G).

So either map the file with the 4G maximum, or (as you say) only map the current file size and remap after growth. I agree that the first one makes sense. But in neither case should the hi redzone size be involved.

Reply:
> The size of the hi redzone is platform-specific...

So `wasm-c-api` should expose more info?

Right now we only have:
```
auto memory_redzone_size_hi(size_t size) -> size_t {
  return V8_TARGET_ARCH_64_BIT ? v8::internal::wasm::kWasmMaxHeapOffset - size : 0;
}
```
but we'd need something like `auto memory_max_heap_size() -> size_t`?

Then:
```
  auto result = mmap(data, wasm::v8::Memory::max_heap_size(),
                     PROT_NONE, MAP_FIXED | MAP_SHARED, fd, 0);
```

Reply:
Not needed, that size is 4G, that's fixed by the Wasm spec. (Until Wasm adds 64 bit memory, but that will need new API anyway.)